src/System/Wlog/IOLogger.hs

```diff
@@ -156,63 +160,67 @@ logMCond logname sev msg cond = do
+          let newlt = createLoggers liTree
           let result = fromJust $ M.lookup lname newlt
           return (LogInternalState newlt liPrefix, result)
   where
-    createLoggers :: [LoggerName] -> LogTree -> LogTree
-    createLoggers xs lt = flipfoldl' addLoggerToTree lt xs -- Add logger to tree
-
-    addLoggerToTree ::  LoggerName -> LogTree ->LogTree
-    addLoggerToTree x lt =
-        if M.member x lt
-            then lt
-            else M.insert x (defaultLogger & lName .~ x) lt
+    createLoggers :: LogTree -> LogTree
+    createLoggers lt = foldl' addLoggerToTree lt (componentsOfName lname)
 
-    defaultLogger :: Logger
-    defaultLogger = Logger Nothing [] (error "log-warper has some strange code") -- ???!??!
+    addLoggerToTree ::  LogTree -> LoggerName -> LogTree
+    addLoggerToTree lt x = if M.member x lt then lt else M.insert x (Logger Nothing [] x) lt
 
 -- | Returns the root logger.
 getRootLogger :: MonadIO m => m Logger
 getRootLogger = getLogger rootLoggerName
 
 -- | Handle a log request.
-handle :: MonadIO m => Logger -> LogRecord -> (LogHandlerTag -> Bool) -> m ()
-handle l lrecord@(LR sev _) handlerFilter = do
-    lp <- getLoggerSeverities nm
-    when (sev `Set.member` lp) $ do
-        ph <- concatMap (view lHandlers) <$> parentLoggers nm
-        forM_ ph $ callHandler lrecord nm
+--
+-- 1. Find the deepest logger that has non-zero handlers to handle log message.
+-- 2. Validate if message severity matches this logger severity
+-- 3. Handle it by all parent handlers.
+handle :: forall m. MonadIO m => Logger -> LogRecord -> (LogHandlerTag -> Bool) -> m ()
+handle l lr@(LR sev _) handlerFilter =
+    traverseAndLog False =<< parentLoggers nm
   where
     nm :: LoggerName
     nm = view lName l
 
-    parentLoggers :: MonadIO m => LoggerName -> m [Logger]
-    parentLoggers = mapM getLogger . componentsOfName
-
-    -- Get the severity we should use. Find the first logger in the
-    -- tree, starting here, with a set severity. If even root doesn't
-    -- have one, assume "Debug".
-    getLoggerSeverities :: MonadIO m => LoggerName -> m Severities
-    getLoggerSeverities name = do
-        pl <- parentLoggers name
-        case mapMaybe (view lLevel) (l : pl) of
-            []    -> pure debugPlus
-            (x:_) -> pure x
-
-    callHandler :: MonadIO m => LogRecord -> LoggerName -> HandlerT -> m ()
-    callHandler lr loggername (HandlerT x) =
+    -- Returns all loggers, root logger last
+    parentLoggers :: LoggerName -> m [Logger]
+    parentLoggers = fmap reverse . mapM getLogger . componentsOfName
+
+    -- Tries to log the message into handlers. sevFiltPassed variable
+    -- denotes the "has log message passed through the first severity
+    -- filter". We only apply severity filter once, the first time we
+    -- encounter it.
+    traverseAndLog :: Bool -> [Logger] -> m ()
+    traverseAndLog _ [] = pass
+    traverseAndLog sevFiltPassed (x:xs) = do
```

Comment:
[`whenNotNull`](https://github.com/serokell/universum/blob/eec7d0dddd8ead098f6b177c2222a3d2475b9119/src/Universum/List/Safe.hs#L64-L66) from `universum` can be used here